pkg/plugins/document/firestore/firestore.go

```diff
@@ -82,13 +82,53 @@ func (s *FirestoreDocService) Delete(key *document.Key) error {
 
 	doc := s.getDocRef(key)
 
+	// Delete any sub collection documents
+	collsIterator := doc.Collections(s.context)
+	for {
+		subColRef, err := collsIterator.Next()
+		if err == iterator.Done {
+			break
+		}
+		if err != nil {
+			return fmt.Errorf("error deleting value: %v", err)
+		}
+
+		for {
+			docsIterator := subColRef.Limit(100).Documents(s.context)
+			numDeleted := 0
+
+			batch := s.client.Batch()
+			for {
+				subDoc, err := docsIterator.Next()
+				if err == iterator.Done {
+					break
+				}
+				if err != nil {
+					return err
+				}
+
+				batch.Delete(subDoc.Ref)
+				numDeleted++
+			}
+
+			// If no more to delete, completed
+			if numDeleted == 0 {
+				break
+			}
+
+			_, err := batch.Commit(s.context)
+			if err != nil {
+				return err
+			}
+		}
```

Comment:
The second inner loop here might be redundant, let me know if I'm missing something but it seems equivalent to:
```go
	collsIterator := doc.Collections(s.context)
	for {
		subColRef, err := collsIterator.Next()
		if err == iterator.Done {
			break
		}
		if err != nil {
			return fmt.Errorf("error deleting value: %v", err)
		}

		docsIterator := subColRef.Limit(100).Documents(s.context)
		numDeleted := 0

		batch := s.client.Batch()
		for {
			subDoc, err := docsIterator.Next()
			if err == iterator.Done {
				break
			}
			if err != nil {
				return err
			}

			batch.Delete(subDoc.Ref)
			numDeleted++
		}

		// If no more to delete, completed
		if numDeleted == 0 {
		        // replace break with continue to restart the outer loop
			continue
		}

		_, err := batch.Commit(s.context)
		if err != nil {
			return err
		}
	}
```
Assuming the break was to skip an empty commit.

Reply:
The reason for the additional loop is there is a batch limit on the number of items which can be deleted (maximum is 500). I have updated this block of code to make it more explicit what it is doing and increased the batching size.

This code was derived from GCP doco examples: https://github.com/GoogleCloudPlatform/golang-samples/blob/09368dd82c1ebc3a8c43e258abc6c1fd9e33277f/firestore/firestore_snippets/save.go#L271-L308

Reply:
cool, clearer with the changes, I'd still recommend including the scope in each for loop per Jye's recommendation above to make it clearer.